mantidproject/mantid
# Migrate from obselete QGLWidget to QOpenGLWidget

**Description of work.**

QGLWidget has not been formally deprecated but there are several issues with it and the docs
mark it has obselete.

On macOS it also seems impossible to correct deal with moving an OpenGL window between
monitors with different pixel densities, while the new widget seems to deal with it seamlessly.
We have removed almost all glViewport calls as the base widget now handles this for us.

It has been necessary to set the default OpenGL profile to compatability mode as the current
code uses older OpenGL api calls. Future work will need to tackle this issue as it is outside of
the scope. A note has been added to the maintenance board.

Provides an alternate and better solution to #35058.



**To test:**

This needs to be tested on a mac with retina screen and an additional lower-resolution screen connected.
Use the following script to determine this information: 

```python
from qtpy.QtWidgets import QApplication
for screen in QApplication.screens():
    print(screen.name())
    print(screen.devicePixelRatio())
```

It should print 2.0 for a retina display and 1.0 for a standard external screen.
If all values are the same then the test of moving the window between screens will not be valid.

All tests will need a workspace with an instrument. Use the following script to create a test data set from
the training course data:

```python
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np

SXD23767 = Load(Filename='SXD23767.raw', LoadMonitors='Exclude')
peaksws = Load('peaks_qLab.nxs')
FindUBUsingFFT(PeaksWorkspace='peaksws', MinD=0.8, MaxD=10)
IndexPeaks(peaksws, Tolerance=0.05)
```

**1. Open on Retina Screen**

* Move workbench to the retina screen
* Click show instrument and check the instrument image is full size in the viewport
* Change to one of the unwrapped views, i.e `CylinderY` and the image should still be fullsize in the viewport

**2. Open on the Standard Def Screen**

* Close any instrument view windows
* Move workbench to the standard definition screen
* Click show instrument and check the instrument image is full size in the viewport
* Change to one of the unwrapped views, i.e `CylinderY` and the image should still be fullsize in the viewport

**3. Move Window Between Screens**

* Close any instrument view windows
* Click show instrument and check the instrument image is full size in the viewport
* Drag the window to the retina screen and check that one the drag is complete the image is still full size in the viewport
* Drag the window back and check again.
* Do the above for the unwrapped view too.

**Zooming**

* When on the Render tab use the mouse to select a rectangle and ensure the zoom takes you to the expected place.

**Overlay peaks workspace**

* Close an instrument view windows
* Load the SXD data set from the training course data
* Also load the `peaks_qLab.nxs` from the training course data
* Show the instrument on the data workspace
* Drag and drop the peaks workspace onto the instrument
* In the "Display Settings" alternate between "Use OpenGL" checked and unchecked
* The peaks should be in the same positions.

**Pick Tab**

* Move to the Pick tab
* Move the mouse over the display and check the miniplot updates with the spectrum values
* Move the mouse over peaks that are labeled and check the miniplot displays the HKL values on the miniplot
* Select the peak picking tool
  * click on a peak on the instrument
  * click at the top of the peak that appears frozen in the miniplot and a new table workspace should have the expected peak values (check the values match those with OpenGL disabled) 

Fixes #34081 



---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
